keps/0030-immutable-parameters.md

```diff
@@ -0,0 +1,173 @@
+
+### Goals
+
+Make it possible for an operator developer to specify that a parameter can not be changed after the installation
+
+### Non-Goals
+
+- Allow modification of immutable parameters after the initial deploy plan is finished
+- Constants that can be used in templating but can not be changed by the user. Immutable parameters can always be set by the user on installation.
+- Immutable Parameters that can be set after the installation - There is a use case for parameters that have an undefined state until they are first used, for example in a specific plan, but can not be changed afterwards. This KEP requires a parameter to receive a value at installation time, either by using a default or provided by the user.
+
+## Proposal
+
+An `immutable` property is added to the parameter definition. If that property is true, the parameter needs to either have a default value or has to be explicitly provided by the user when the operator is installed. The parameter can not be updated with `kudo update -p` anymore.
+
+```yaml
+parameters:
+  - name: NUM_TOKENS
+    description: The number of tokens on each node of the cluster
+    default: 256
+    immutable: true
+```
+
+The default value for `immutable` is `false`.
+
+### User Stories
+
+- [#652](https://github.com/kudobuilder/kudo/issues/652) Make it possible to specify immutable parameter
+
+- NUM_TOKENS in the Cassandra operator
+- STORAGE_CLASS in the Zookeeper operator
+- DISK_SIZE, STORAGE_CLASS and PERSISTENT_STORAGE in the Kafka operator
+
+### Implementation Details/Notes/Constraints
+
+#### Installation
+- If a parameter definition is immutable, it must either have a default value or needs to be marked as required.
+- If no value for an immutable parameter is given on installation, KUDO copies the default value from the OperatorVersion into the instance. This makes sure that the parameter value will never change, even if instance is changed to use a newer OperatorVersion that has different defaults.
+
+#### Updates
+- KUDO CLI will not perform any checks with regards to immutability on installation or updates
+- The admission webhook verifies if a parameter is immutable
+  - If the value of any immutable parameter changes from the current value it rejects the whole update
+  - If the value of all immutable parameters are the same as their current values it allows the update
+
+#### Upgrades
+- Upgrades to a new OperatorVersions:
+  - A parameter can be made immutable in a newly released OperatorVersion.
+    - When executing `kudo upgrade` the user has to explicitly set the value for parameters that are changed from mutable to immutable
+    - If an parameter is changed from mutable to immutable and the user does not explicitly specifies a parameter the upgrade will abort with an error message
+    - This ensures that the user knows about the new immutability
```

Comment:
For upgrades, we could silently use the current value, but I think it's good practice to get an informed user decision. This case shouldn't happen too often and therefore I think it's ok to make it a bit harder to use.